arduino/discovery/discoverymanager/discoverymanager.go

```diff
@@ -60,234 +79,174 @@ func (dm *DiscoveryManager) IDs() []string {
+// The watcher must be closed when it is no longer needed with the Close method.
+func (dm *DiscoveryManager) Watch() (*PortWatcher, error) {
+	dm.Start()
 
-		if err := d.Stop(); err != nil {
-			dm.remove(d.GetID())
-			return fmt.Errorf(tr("stopping discovery %[1]s: %[2]w"), d.GetID(), err)
+	watcher := &PortWatcher{
+		feed: make(chan *discovery.Event, 10),
+	}
+	watcher.closeCB = func() {
+		dm.watchersMutex.Lock()
+		defer dm.watchersMutex.Unlock()
+		delete(dm.watchers, watcher)
+		close(watcher.feed)
+	}
+	go func() {
+		dm.watchersMutex.Lock()
+		// When a watcher is started, send all the current active ports first...
+		for _, cache := range dm.watchersCache {
+			for _, ev := range cache {
+				watcher.feed <- ev
+			}
 		}
-		return nil
-	})
+		// ...and after that add the watcher to the list of watchers receiving events
+		dm.watchers[watcher] = true
+		dm.watchersMutex.Unlock()
+	}()
+	return watcher, nil
 }
 
-// QuitAll quits all the discoveries managed by this DiscoveryManager.
-// Returns an error for each discovery that fails quitting
-func (dm *DiscoveryManager) QuitAll() []error {
-	errs := dm.parallelize(func(d *discovery.PluggableDiscovery) error {
-		if d.State() == discovery.Dead {
-			// Stop! Stop! It's already dead!
-			return nil
+func (dm *DiscoveryManager) startDiscovery(d *discovery.PluggableDiscovery) (discErr error) {
+	defer func() {
+		// If this function returns an error log it
+		if discErr != nil {
+			logrus.Errorf("Discovery %s failed to run: %s", d.GetID(), discErr)
 		}
+	}()
 
-		d.Quit()
-		return nil
-	})
-	return errs
-}
-
-// List returns a list of available ports detected from all discoveries
-// and a list of errors for those discoveries that returned one.
-func (dm *DiscoveryManager) List() ([]*discovery.Port, []error) {
-	var wg sync.WaitGroup
-	// Use this struct to avoid the need of two separate
-	// channels for ports and errors.
-	type listMsg struct {
-		Err  error
-		Port *discovery.Port
+	if err := d.Run(); err != nil {
+		return fmt.Errorf(tr("discovery %[1]s process not started: %[2]w"), d.GetID(), err)
 	}
-	msgChan := make(chan listMsg)
-	dm.discoveriesMutex.Lock()
-	discoveries := []*discovery.PluggableDiscovery{}
-	for _, d := range dm.discoveries {
-		discoveries = append(discoveries, d)
-	}
-	dm.discoveriesMutex.Unlock()
-	for _, d := range discoveries {
-		wg.Add(1)
-		go func(d *discovery.PluggableDiscovery) {
-			defer wg.Done()
-			if d.State() != discovery.Running {
-				// Discovery is not running, it won't return anything
-				return
-			}
-			ports, err := d.List()
-			if err != nil {
-				msgChan <- listMsg{Err: fmt.Errorf(tr("listing ports from discovery %[1]s: %[2]w"), d.GetID(), err)}
-			}
-			for _, p := range ports {
-				msgChan <- listMsg{Port: p}
-			}
-		}(d)
+	eventCh, err := d.StartSync(5)
+	if err != nil {
+		return fmt.Errorf(tr("start syncing discovery %[1]s: %[2]w"), d.GetID(), err)
 	}
 
+	// XXX do better cleanup if the discovery fails to start
```

Comment:
Needs to be addressed?

Reply:
When I wrote this, I was thinking about removing the failing discovery from the discoverymanager list, but looking at the code a bit more now I think it's not needed, so I'll just remove the comment.